src/js/components/DSLForm.js

```diff
@@ -0,0 +1,46 @@
+import React, {PropTypes} from 'react';
+
+import DSLExpression from '../structs/DSLExpression';
+
+/**
+ * This component wraps one or more DSLOptionSection components and
+ * handles the transparent mutation of the expression based on user actions.
+ */
+class DSLForm extends React.Component {
+  /**
+   * @override
+   */
+  render() {
+    const {expression, onApply, onChange, sections} = this.props;
+
+    // Render each group component wrapped with a dedicated form that receives
+    // the updates targeting each dedicated component.
+    return (
+      <div>
+        {sections.map((SectionComponent, key) => {
```

Comment:
I had a dilemma for what to separate here.

* If I separate the `getComponent` function I will have to add an unnecessary overhead for binding the function, plus the function will only return a component.
* If I separate the `getComponents` function, as the entire render section becomes just a `<div>` 

So eventually I thought that the work that function does is quite small regarding the overhead of creating a function. But if you want to move the logic from within the JSX block, it might actually make sense to populate a variable `sections` right before. I might do that.